### PR TITLE
Dont hold onto references of saved tensors in backward

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88248
* __->__ #88247
* #88208

This improves memory compression of resnet18 on inductor non-cudagraphs from .78 -> .0.84.

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -472,12 +472,16 @@ def forward(ctx, *deduped_flat_tensor_args):
...
             all_args = list(ctx.symints) + list(ctx.saved_tensors) + list(contiguous_args)
+            del contiguous_args
             if CompiledFunction.compiled_bw is None:
+                mode = torch._subclasses.fake_tensor.FakeTensorMode()
+                fake_args = [mode.from_tensor(x) if isinstance(x, torch.Tensor) else x for x in all_args]
                 context = disable_autocast_manager if disable_amp else nullcontext
                 with context(), track_graph_compiling("backward", True):
                     CompiledFunction.compiled_bw = aot_config.bw_compiler(
-                        bw_module, all_args
+                        bw_module, fake_args
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:482]: @Chillee how come the backward compiler needs fake args, didn't we already synthesize then when creating the joint graph?

**Review by ezyang — APPROVED** [review]
Nice catch. Any chance we can test it?

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (Chillee) [reply functorch/_src/aot_autograd.py:482]: What do you mean? We created them, but in the future, it'd be possible for them to be other tensor subclasses.

Also, the other issue is that need to pass symints from the backwards to the forwards pass.

Comment (ezyang) [reply functorch/_src/aot_autograd.py:482]: It's hard for me to believe this case actually works correctly lol.

Comment (ezyang) [reply functorch/_src/aot_autograd.py:482]: Like, how do you even fakify an arbitrary torch dispatch tensor subclass

**Comment by eellison** [comment]
Filed https://github.com/pytorch/torchdynamo/issues/1829: at the moment this is pretty difficult to test.

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -472,12 +472,15 @@ def forward(ctx, *deduped_flat_tensor_args):
         def backward(ctx, *flat_args):
             contiguous_args = [t.contiguous() if torch.is_tensor(t) else t for t in flat_args]
             all_args = list(ctx.symints) + list(ctx.saved_tensors) + list(contiguous_args)
+            del contiguous_args
             if CompiledFunction.compiled_bw is None:
+                # TODO - pass in fake tensors ?
```
Comment (eellison) [inline functorch/_src/aot_autograd.py:477]: this wasn't actually necessary for memory savings and had a bad interaction with the pr above this + didn't repro with minifier - deferring on passing in fake for now

**Pull request closed without merging**